Adrià Melús - 1527144

# Pràctica 3 d'APC

## 1.Introducció

Elon Musk és considerada la persona amb més diners del planeta actualment; és el director executiu de la empresa aerospacial SpaceX i de la d'automòvils elèctrics Tesla, i fa poc va rebre el títol de la persona del any per part de la revista Time. 

Elon Musk també és algú que destaca per la seva activitat a Twitter, on publica memes o frases que deixen a tot el món pensant si realment les diu en serio. Mai està massa ocupat per comentar alguna broma o un pensament espontani. Tot i això, els seus intents per semblar graciós, per lo general acaben mosrant la seva irresponsabilitat al publicar tantes afirmacions confuses, insensibles o incorrectes.

En honor al seu nou títol, que segons la publicació és otorgat a "la persona o persones que més han afectat les notícies i les nostres vides, cap a bé o cap a malament".

Aquest és l'informe de  l'anàlisi de resultats en aprenentatge computacional d'una base de dades de kaggle que conté més de 25.000 tweets d'Elon Musk: https://www.kaggle.com/taruntiwarihp/elon-musk-tweets

El propòsit principal d'aquest anàlisi és el de predir el sentiment d'un tweet segons sigui positiu o negatiu:

<img src="../figures/positive-negative.jpg">

Following topics would be covered as an outline for this article:

Data cleaning and preparation.
Analysis of the data.
Analysis and preprocessing text with NLTK.
Resampling data.
Model creation to predict sentiment.


### LLibreries

Per aquest projecte haurem de instalar les seguents llibreries:

Matplotlib, seaborn i wordcloud per la visualització gràfica de dades.

Pandas i numpy pel tractament d'informació.

SkLearn per l'entrenament dels nostres models.

NLTK pel processament de text.

imblearn pel "resampling" de la informació.

In [26]:
# Import 
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import string
import scikitplot as skplt
import plotly.graph_objs as go
import nltk

from nltk.tokenize import RegexpTokenizer
from nltk.corpus import stopwords
from nltk.sentiment.vader import SentimentIntensityAnalyzer

from wordcloud import WordCloud, STOPWORDS
from prettytable import PrettyTable
from plotly.offline import iplot

from imblearn.combine import SMOTETomek

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

nltk.download('stopwords', quiet=True);
nltk.download('vader_lexicon', quiet=True);

ModuleNotFoundError: No module named 'scikitplot'

## 2. Informació de les dades

Primer de tot observarem detalladament les dades i les entendrem. A continuació, si és necessari, netejarem el dataset i el prepararem per analitzar-lo.

In [33]:
# Carregar el Dataset

dataset = "../dataset/Elon Musk.csv"
dades = pd.read_csv(dataset)

# Eliminem la primera columna perquè es tracta d'un index que no ens fa falta
dades.drop(dades.columns[0], inplace = True, axis = 1)
dades.head(1)


tweets sentiment
0  @elonmusk @realfuckingnews @business Ancient c...       pos

In [35]:
# Quantes files i columnes hi ha al dataset? Cada fila representa un tweet
print("Hi ha {} columnes".format(dades.shape[1]), "i {} files.".format(dades.shape[0]))
print(dades.dtypes)

Hi ha 2 columnes i 25757 files.
tweets       object
sentiment    object
dtype: object


Com es pot observar, aquest dataset conté 25757 files i 2 columnes. Cada fila correspon a un tweet i inclou dues variables:

tweets: Variable de tipus object amb el contingut d'un tweet que cita @elonmusk.
sentiment: Variable de tipus object que indica el sentiment del tweet. L'atribut sentiment pot prendre dos valors: pos (positiu) o neg (negatiu)

Ara comprovarem si hi ha algun valor null o Nan

In [36]:
null_counts = dades.isnull().sum()
null_counts[null_counts>0]

Series([], dtype: int64)

No s'ha trobat cap valor null o Nan, per tant no farà falta eliminar cap fila, de moment.

## 3. Anàlisi de la informació

### 3.1. (Distribució dels sentiments

## Mètode d'aprenentatge computacional aplicat,

## Resultats

## 6. Conclusions


In [ ]:
## Conclusions